File: app/src/main/java/com/example/androidscript/Activities/ArkKnights/ArkKnightsEditor.java
package com.example.androidscript.Activities.ArkKnights;

import androidx.appcompat.widget.SwitchCompat;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;

import com.example.androidscript.FloatingWidget.FloatingWidgetService;
import com.example.androidscript.Activities.StartService;
import com.example.androidscript.R;
import com.example.androidscript.UITemplate.Editor;
import com.example.androidscript.util.BtnMaker;
import com.example.androidscript.util.DebugMessage;
import com.example.androidscript.util.FileOperation;
import com.example.androidscript.util.ScreenShot;

import java.util.Vector;

import static java.lang.Math.max;
import static java.lang.Math.min;

public class ArkKnightsEditor extends Editor {
    SwitchCompat TillEmpty;
    SwitchCompat EatMedicine;
    SwitchCompat EatStone;
    EditText Repeat;
    private boolean isTillEmpty = false;
    private boolean isEatMedicine = false;
    private boolean isEatStone = false;
    private String SelectedScript;
    private int nRepetition;
    public static final String FolderName = "ArkKnights/";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_ark_knights_editor);
        Repeat = findViewById(R.id.Repetition);
        TillEmpty = findViewById(R.id.tillEmpty);
        TillEmpty.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                Repeat.setVisibility(View.GONE);
            } else {
                Repeat.setVisibility(View.VISIBLE);
            }
        });
        EatMedicine = findViewById(R.id.EatMedicine);
        EatStone = findViewById(R.id.EatStone);
        BtnMaker.registerOnClick(R.id.set_script, this, v -> {
            if(!StartService.IsAuthorized(this)){
                this.startActivity(new Intent(this, StartService.class).putExtra("Orientation", "Landscape"));
            }else{
                CheckState();
                GetRepetition();
                if (isEatStone) {
                    FloatingWidgetService.setScript(FolderName, "AutoFightEat.txt", new String[]{String.valueOf(nRepetition), "PressRestore.png"});
                } else if (isEatMedicine) {
                    FloatingWidgetService.setScript(FolderName, "AutoFightEat.txt", new String[]{String.valueOf(nRepetition), "PressRestoreMedicine.png"});
                } else {
                    FloatingWidgetService.setScript(FolderName, "AutoFight.txt", new String[]{String.valueOf(nRepetition)});
                }
                StartService.StartFloatingWidget(this);
            }
        });
    }

    private void GetRepetition() {
        if (isTillEmpty) {
            nRepetition = 100000;
        } else {
            try {
                nRepetition = Integer.parseInt(Repeat.getText().toString());
            } catch (NumberFormatException e) {
                nRepetition = 0;
            }
        }
    }

    private void CheckState() {
        isTillEmpty = TillEmpty.isChecked();
        isEatMedicine = EatMedicine.isChecked();
        isEatStone = EatStone.isChecked();
    }

    @Override
    public String getFolderName() {
        return FolderName;
    }

    double resizeRatio;

    @Override
    protected void resourceInitialize() {
        try {
            FileOperation.readDir(FolderName);
            String[] allFiles = getAssets().list(FolderName);//List all file
            for (String file : allFiles) {
                    getResource(FolderName, file);
            }
        } catch (Exception e) {
            DebugMessage.printStackTrace(e);
        }
        resizeRatio = min(ScreenShot.getHeight(), ScreenShot.getWidth()) / 1152.0;//ArkKnights seems to be height dominate.
        // resizeRatio = ScreenShot.getWidth() / 2432.0;
        writePress();
        writeTryPress();
    }

    private void writePress() {
        Vector<String> buffer = new Vector<>();
        buffer.add("Tag $Start");
        buffer.add("ClickPic $1 " + resizeRatio);
        buffer.add("Wait $2");
        buffer.add("IfGreater $R 0");
        buffer.add("JumpTo $Start");
        FileOperation.writeLines(FolderName + "Press.txt", buffer);
    }

    private void writeTryPress() {
        Vector<String> buffer = new Vector<>();
        buffer.add("ClickPic $1 " + resizeRatio);
        buffer.add("Return $R");
        FileOperation.writeLines(FolderName + "TryPress.txt", buffer);
    }
}

File: app/src/main/java/com/example/androidscript/activities/arknights/ArkKnightsEditor.kt
package com.example.androidscript.activities.arknights

import android.content.Intent
import com.example.androidscript.util.DebugMessage
import com.example.androidscript.util.FileOperation
import android.os.Bundle
import com.example.androidscript.util.BtnMaker
import com.example.androidscript.R
import com.example.androidscript.floatingwidget.FloatingWidgetService
import android.widget.EditText
import androidx.appcompat.widget.SwitchCompat
import android.widget.CompoundButton
import com.example.androidscript.floatingwidget.ScreenShotService
import com.example.androidscript.activities.StartServiceActivity
import android.view.*
import com.example.androidscript.uitemplate.Editor
import java.lang.Exception
import java.lang.Integer.min
import java.util.*

class ArkKnightsEditor : Editor() {
    private lateinit var tillEmpty: SwitchCompat
    private lateinit var eatMedicine: SwitchCompat
    private lateinit var eatStone: SwitchCompat
    private lateinit var repeat: EditText
    private val isTillEmpty: Boolean
        get() = tillEmpty.isChecked
    private val isEatMedicine: Boolean
        get() = eatMedicine.isChecked
    private val isEatStone: Boolean
        get() = eatStone.isChecked
    private val nRepetition: Int
        get() {
            return when (isTillEmpty) {
                true -> 10000
                false -> repeat.text.toString().toInt()
            }
        }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_ark_knights_editor)
        repeat = findViewById(R.id.Repetition)
        tillEmpty = findViewById(R.id.tillEmpty)
        tillEmpty.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            if (isChecked) {
                repeat.visibility = View.GONE
            } else {
                repeat.visibility = View.VISIBLE
            }
        }
        eatMedicine = findViewById(R.id.EatMedicine)
        eatStone = findViewById(R.id.EatStone)
        BtnMaker.registerOnClick(R.id.set_script, this) {
            if (!StartServiceActivity.isAuthorized(this)) {
                this.startActivity(
                    Intent(this, StartServiceActivity::class.java).putExtra(
                        "Orientation",
                        "Landscape"
                    )
                )
            } else {
                if (isEatStone) {
                    FloatingWidgetService.setScript(
                        folderName,
                        "AutoFightEat.txt",
                        arrayOf(nRepetition.toString(), "PressRestore.png")
                    )
                } else if (isEatMedicine) {
                    FloatingWidgetService.setScript(
                        folderName,
                        "AutoFightEat.txt",
                        arrayOf(nRepetition.toString(), "PressRestoreMedicine.png")
                    )
                } else {
                    FloatingWidgetService.setScript(
                        folderName,
                        "AutoFight.txt",
                        arrayOf(nRepetition.toString())
                    )
                }
                StartServiceActivity.startFloatingWidget(this)
            }
        }
    }


    private var resizeRatio = 0.0
    override fun resourceInitialize() {
        try {
            FileOperation.readDir(folderName)
            val allFiles = assets.list(folderName) //List all file
            for (file in allFiles!!) {
                getResource(folderName, file!!)
            }
        } catch (e: Exception) {
            DebugMessage.printStackTrace(e)
        }
        resizeRatio = min(
            ScreenShotService.height,
            ScreenShotService.width
        ) / 1152.0 //ArkKnights seems to be height dominate.
        // resizeRatio = ScreenShot.getWidth() / 2432.0;
        writePress()
        writeTryPress()
    }

    private fun writePress() {
        val buffer = Vector<String>()
        buffer.add("Tag \$Start")
        buffer.add("ClickPic $1 $resizeRatio")
        buffer.add("Wait $2")
        buffer.add("IfGreater \$R 0")
        buffer.add("JumpTo \$Start")
        FileOperation.writeLines(folderName + "Press.txt", buffer)
    }

    private fun writeTryPress() {
        val buffer = Vector<String>()
        buffer.add("ClickPic $1 $resizeRatio")
        buffer.add("Return \$R")
        FileOperation.writeLines(folderName + "TryPress.txt", buffer)
    }

    companion object {
        const val folderName = "ArkKnights/"
    }
}